respy/simulate.py

```diff
@@ -446,3 +508,86 @@ def _random_choice(choices, probabilities):
         choices = np.arange(choices)
 
     return choices[indices]
+
+
+def apply_law_of_motion(choice, states, period, optim_paras):
+    """Apply the law of motion to get the states in the next period.
+
+    This function changes experiences and previous choices according to the choice in
+    the current period, to get the states of the next period.
+
+    Parameters
+    ----------
+    choice : np.ndarray
+        Array with shape (n_individuals,) containing the current choice.
+    states : np.ndarray
+        Array with shape (n_individuals, n_state_space_dim) containing the state of each
+        individual.
+
+    """
+    n_individuals = states.shape[0]
+    n_choices_w_exp = len(optim_paras["choices_w_exp"])
+    n_lagged_choices = optim_paras["n_lagged_choices"]
+
+    # Update work experiences.
+    states[np.arange(n_individuals), choice] = np.where(
+        choice < n_choices_w_exp,
+        states[np.arange(n_individuals), choice] + 1,
+        states[np.arange(n_individuals), choice],
+    )
+
+    # Update lagged choices by shifting all lags by one and inserting choice in
+    # the first position.
+    if n_lagged_choices:
+        states[:, n_choices_w_exp + 1 : n_choices_w_exp + n_lagged_choices] = states[
+            :, n_choices_w_exp : n_choices_w_exp + n_lagged_choices - 1
+        ]
+        states[:, n_choices_w_exp] = choice
+
+    state_space_columns = _create_state_space_columns(optim_paras)
```

Comment:
Since we have to create the same container every period would it make sense to just update the df every period and only  create it once before the loop?

Reply:
Very good comment. The function now works with the DataFrame instead of NumPy arrays which makes the transition to the next period more explicit and easier to understand.

I also tried to avoid copies and calculation as much as possible and instead used renaming.